pkg/restore/service_action.go

```diff
@@ -52,15 +61,49 @@ func (a *serviceAction) Execute(obj runtime.Unstructured, restore *api.Restore)
 		delete(spec, "clusterIP")
 	}
 
+	preservedPorts, err := getPreservedPorts(obj)
+	if err != nil {
+		return nil, nil, err
+	}
+
 	ports, err := collections.GetSlice(obj.UnstructuredContent(), "spec.ports")
 	if err != nil {
 		return nil, nil, err
 	}
 
 	for _, port := range ports {
 		p := port.(map[string]interface{})
+		var name string
+		if nameVal, ok := p["name"]; ok {
+			name = nameVal.(string)
+		}
+		if preservedPorts[name] {
+			continue
+		}
 		delete(p, "nodePort")
 	}
 
 	return obj, nil, nil
 }
+
+func getPreservedPorts(obj runtime.Unstructured) (map[string]bool, error) {
+	preservedPorts := map[string]bool{}
+	if collections.Exists(obj.UnstructuredContent(), metadataAnnotPath) {
+		annot, err := collections.GetMap(obj.UnstructuredContent(), metadataAnnotPath)
+		if err != nil {
+			panic(fmt.Errorf("%q not available after existence check", metadataAnnotPath))
```

Comment:
I think using the `serviceAction.log` field would be better than a panic here.

Reply:
To elaborate my decision: I chose to panic to express the invariant that we can only hit an error in case of a logic bug in the code (we established that the metadata annotation must exist a few lines above).

I'm happy to handle the error differently, but then the question is how to deal with the situation? We can either log and leave the map of preserved node ports empty (causing any found node ports to be deleted by Ark, possibly leading to unexpected behavior for the user) or propagate the error up the stack (which I suppose would make the restore fail partially or completely). Personally, I'm leaning towards the latter so that the error is surfaced and not worked around implicitly. WDYT?

Reply:
Please don't panic from an item action, as these typically run as plugins. We have an in-flight PR to handle plugins exiting (for whatever reason), but as for what's in master right now, if the plugin process exits/crashes, Ark won't restart it.

We are also intending to switch from unstructured to typed objects in places such as this (i.e. convert to a `v1.Service`). If the conversion fails, we'll return an error. Otherwise, we'll be able to work with things like `service.Annotations` without worrying if the map exists in the object.

Reply:
Thanks for the extra info. I changed the code to return an error.